### ISSUE TITLE
Cred data for analysis in output file

### ISSUE THREAD

**Issue by s-ben** [sourcecred/sourcecred#2949]
**Problem**
In TimelineCred, basic graph data was output to `credResult.json`. I was using data in this file to do analysis in [Python](https://github.com/s-ben/sourcecred-analysis), for instance to calculate Cred stats for updates for the Maker community. In CredRank, due to virtualization of some of the data, we now have `credGraph.json.gzip`, which contains some data, but not enough to do meaningful analysis. Myself -and likely others- are going to need more data in order to perform meaningful analysis in a number of use cases, such as:

- General data viz of graph
- Analysis of weight changes
- Gaming detection
- Research

We have some observable notebooks that pull data from the graph into consumable formats. E.g. the [Cred and Grain CSV Downloader](https://observablehq.com/@sourcecred/cred-and-grain-csv-downloader?branchUrl=https://github.com/sourcecred/cred/tree/gh-pages). This is great. But I imagine a demand from myself and others for lower-level access to more data, which does not require knowledge of Javascript or the relevant APIs. Already, other communities have created their own Python scripts for doing data visualization and automation, which have likely broken with CredRank. 

**Describe the solution you'd like**

I would like lower-level graph data in a more easily consumable format, such as CSV or JSON file. Perhaps a command on the CLI that optionally creates the file, to avoid bloat? No ideas on the data schema, as don't have experience there.. 

![image](https://user-images.githubusercontent.com/1634777/114824649-104d3b80-9d7a-11eb-8349-d21c9cb27dec.png)

Jk. Kinda....I can imagine people (myself included) wanting granular Cred data for analysis. Below I give the MVP (basic data that would unblock me) and additional data we'll have demand for eventually. 

*Basic data*

- Data for each node (users and contributions), including: Cred score, Cred minted, author, platform, node type, timestamp, description.
- Cred data over time, including user Cred scores and contribution Cred scores (if relevant in CredRank). I.e. I want to be able to create plots of a user or contribution's Cred score over time. 
- Data for each edge, such as edge type (what's in `address` might do?), reversed, source, destination, transition probability, Cred flow. 

*Nice to have (but will need in future)*

- Node metadata, such as the number and type of emoji reactions on a post in Discord, category or tags applied to a topic in Discourse, or any other metadata relevant enough to fetch and include in the database. This will be important, in particular as people try to categorize Cred. For instance, trying to identify "design Cred", or cluster emojis related to care work. Have heard people float many ideas that depend on the ability to do this more fine-grained analysis. I almost want to include category and tags for Discourse topics in basic data, as it would make my analysis of Maker's "governance boost" implementation using category and tags minting much easier (currently using PITA workaround). 

**Describe alternatives you've considered**
We have a number of observable notebooks now that pull graph data and output to CSV or visualize it:
- [Contribution Explorer Beta](https://observablehq.com/embed/@sourcecred/credgraph-contribution-explorer-beta?cell=title&cell=viewof+userPicker&cell=viewof+intervalPicker&cell=credTable&cell=total)
- [Cred and Grain csv downloader](https://observablehq.com/@sourcecred/cred-and-grain-csv-downloader )
- [Nodes and edges Csv downloader](https://observablehq.com/@sourcecred/nodes-and-edges-csv-downloader)

These may work for now, but still require a bit of parsing and cross-referencing to get some basic data (e.g. timestamps). 

**Additional context**

Long-term, I think we'll have numerous consumers of this data, many of whom will not have the ability to pull data from the existing APIs in JS. So should be a good investment IMHO.

**Comment by blueridger** [comment]
A new CLI command and main API would be ideal.

> Data for each node (users and contributions), including: Cred score, Cred minted, author, platform, node type, timestamp, description.

Can start with outputting credGraph.nodes() and credGraph.edges(). Platform and node type can be easily parsed by the client using the address, but we could also upgrade the output to add a field for ease of access. Author is a little trickier, and is probably best left to the client for parsing from either address or connected nodes. For timestamp, we should probably load the weightedGraphs and cross reference the addresses, or maybe it'd be worth actually persisting that through into the CredGraph.

> Cred data over time, including user Cred scores and contribution Cred scores (if relevant in CredRank). I.e. I want to be able to create plots of a user or contribution's Cred score over time.

We have plans to output user Cred scores over time here https://github.com/sourcecred/sourcecred/issues/2859

Contribution cred scores over time isn't supported.

> Data for each edge, such as edge type (what's in address might do?), reversed, source, destination, transition probability, Cred flow.

credGraph.edges() should work for all of this, except Cred Flow, which will need to be added during output upgrade.

**Comment by s-ben** [comment]
> Can start with outputting credGraph.nodes() and credGraph.edges(). Platform and node type can be easily parsed by the client using the address, but we could also upgrade the output to add a field for ease of access.

Parsing from address type is relatively simple. However, we might as well include that logic. Otherwise that logic will have to get rewritten over and over by each consumer of this data. 

> Author is a little trickier, and is probably best left to the client for parsing from either address or connected nodes.

Does this need to be tricky? In my python scripts for instance, I rank the top Cred-earning posts per week. This involves filtering the contribution nodes, then parsing the nodes. Below is an example node:

```
{'address': ['sourcecred',
  'discourse',
  'post',
  'https://forum.makerdao.com',
  '30091'],
 'cred': 88.39602087108804,
 'credOverTime': [],
 'description': '[#55 on [Informal Poll] - Increasing the SourceCred payments](https://forum.makerdao.com/t/7489/55)',
 'timestamp': 1618732674894,
 'user': ''}
 ```

Since there's no link to the user, I am currently doing this manually. I click the link in the description into a browser, then copy/paste the username into a markdown table. I suppose I could look at the edge data (there has to be some link to follow there). But this seems like a lot of work for analysts to go through. And I imagine filtering contribution nodes, then looking at the authors of those nodes will be fairly common. As will other analysis paths that link contributions to authors. Might be worth coding the logic here to create the author variable? 

> For timestamp, we should probably load the weightedGraphs and cross reference the addresses, or maybe it'd be worth actually persisting that through into the CredGraph.

No idea whether persisting through the CredGraph makes sense, but from an analysis perspective, the timestamp of a contribution is a basic element of analysis. Having to recreate a timeline from relative epoch timestamps seems like something better done here than everyone having to re-implement that logic themselves. 

> Contribution cred scores over time isn't supported.

So back in the day (like, early days), they were supported. And I found it useful. For instance, I could plot the exponential decay of a contribution's Cred over time (which is unintuitive for most users --or is that decay gone in CredRank?)). It's also useful to see how a contribution is revalued over time. Currently, we don't really do much historical Cred rebalancing. But it could be useful for instance to plot how an older contribution grew in importance as it got used. This could also be useful for detecting gaming. One possible attack is revaluing old contributions to boost your score. Which likely will go undetected by users through the main UIs. Systematic detection of some kind will be needed. 

I think it was removed for performance reasons? Not necessary I suppose if it's a lot of effort, but I think it's an assumption worth challenging.

**Comment by blueridger** [comment]
Author is tricky because it is defined at the plugin level and might not be generalizable across different plugins. However, plugins do have an identity getter, so we could just start with identity nodes and build a list of nodes that flow into identities. That'd be doable and I think meet your needs.

It'd be pretty simple to persist node timestamps / node epoch timestamps into the markovProcessGraph, that seems like a good thing to do.

CredRank does not have a concept of a contribution's cred over time. Contributions are only linked to 1 epoch. The decay over time is not a feature of credrank. In order to understand cred over time, you'd have to compare credgraphs at multiple points in a git history, which is beyond the scope of what our package does, but is an interesting problem area to someday tackle.

**Comment by s-ben** [comment]
Sounds good. A couple clarifying questions:

> Author is tricky because it is defined at the plugin level and might not be generalizable across different plugins. However, plugins do have an identity getter, so we could just start with identity nodes and build a list of nodes that flow into identities. That'd be doable and I think meet your needs.

That should work. I assume the output of this would be a user field/column for each contribution? Or just a separate list of nodes and associated identities that I can cross reference with node id or something?

> It'd be pretty simple to persist node timestamps / node epoch timestamps into the markovProcessGraph, that seems like a good thing to do.

Does node timestamps here mean the absolute timestamps, or would the absolute timestamps still need to be calculated as node epoch timestamps + intra-node-timestamp?
